libs/stats/odc/stats/io.py

```diff
@@ -242,5 +284,156 @@ def dump(self, task: Task, ds: Dataset, aux: Optional[Dataset] = None) -> Delaye
+
+        paths = task.paths("absolute", ext=self._band_ext)
+        cogs = self._ds_to_cog(ds, paths)
+
+        if aux is not None:
+            aux_paths = {
+                k: task.aux_path(k, relative_to="absolute", ext=self._band_ext)
+                for k in aux.data_vars
+            }
+            cogs.extend(self._ds_to_cog(aux, aux_paths))
+
+        thumbnail_cogs = []
+
+        input_geobox = GridSpec(shape=task.geobox.shape, 
+                                transform=task.geobox.transform, 
+                                crs=CRS.from_epsg(task.geobox.crs.to_epsg()))
+
+        # if we do not define the r/g/b values, we will grab the first avaialble variable shape to 
+        # generate full zero numpy
+        zero_band = numpy.zeros_like(ds[list(ds.keys())[0]].values.reshape([task.geobox.shape[0], 
+                                                                            task.geobox.shape[1]]))
+
+        if task.product.preview_image_singleband:
+            # single_band: {"measurement": "count_clear", "lookup_table": {"0":[150,150,110]}}, or
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['blue']} or 
+            # single_band: {"measurement":'frequency', 'bit':1, 'display_bands': ['red', 'green', 'blue']}
+            for single_band in task.product.preview_image_singleband:
+                band = single_band['measurement']
+                thumbnail_path = odc_file_path.split('.')[0] + f"_{band}_thumbnail.jpg"
+
+                tuning_pixels, stretch = FileWrite()._filter_singleband_data(data=ds[band].values.reshape([task.geobox.shape[0], 
+                                                                                                           task.geobox.shape[1]]),
+                                                                             bit=single_band['bit'] if 'bit' in single_band else None,
+                                                                             lookup_table=single_band['lookup_table'] if 'lookup_table' in single_band else None)
+                if 'bit' in single_band:
+                    # the bit operation return tuning_pixels is numapy.array
+                    if 'display_bands' not in single_band: # if no display_band, duplicate it as r,g,b
+                        tuning_pixels = [tuning_pixels, tuning_pixels, tuning_pixels]
+                    else:
+                        display_pixels = []
+                        for display_band in ["red", "green", "blue"]:
+                            display_pixels.append(tuning_pixels) if display_band in single_band["display_bands"] else display_pixels.append(zero_band)
+                        tuning_pixels = display_pixels # make sure the tuning_pixcels format is [numpy.array, numpy.array, numpy.array] == [r, g, b]
+
+                thumbnail_bytes = FileWrite().create_thumbnail_from_numpy(rgb=tuning_pixels,
```

Comment:
I think preview management code should be taken out of this method, it's already way too long, makes it hard to reconstruct data flows.